src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -40,32 +42,52 @@ internal static class DocumentBasedFixAllProviderHelpers
         var workItemCount = fixAllKind == FixAllKind.CodeFix ? 3 : 2;
         progressTracker.AddItems(fixAllContexts.Length * workItemCount);
 
-        // Process each context one at a time, allowing us to dump any information we computed for each once done with it.
-        var currentSolution = solution;
-        foreach (var fixAllContext in fixAllContexts)
+        using var _1 = PooledDictionary<DocumentId, (SyntaxNode? node, SourceText? text)>.GetInstance(out var allContextsDocIdToNewRootOrText);
         {
-            Contract.ThrowIfFalse(fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project
-                or FixAllScope.ContainingMember or FixAllScope.ContainingType);
-            currentSolution = await FixSingleContextAsync(currentSolution, fixAllContext, progressTracker, getFixedDocumentsAsync).ConfigureAwait(false);
-        }
+            // First, iterate over all contexts, and collect all the changes for each of them.  We'll be making a lot of
+            // calls to the remote server to compute diagnostics and changes.  So keep a single connection alive to it
+            // so we never resync or recompute anything.
+            using var _2 = await RemoteKeepAliveSession.CreateAsync(solution, originalFixAllContext.CancellationToken).ConfigureAwait(false);
 
-        return currentSolution;
-    }
+            foreach (var fixAllContext in fixAllContexts)
+            {
+                Contract.ThrowIfFalse(
+                    fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project or FixAllScope.ContainingMember or FixAllScope.ContainingType);
 
-    private static async Task<Solution> FixSingleContextAsync<TFixAllContext>(
-        Solution currentSolution,
-        TFixAllContext fixAllContext,
-        IProgress<CodeAnalysisProgress> progressTracker,
-        Func<TFixAllContext, IProgress<CodeAnalysisProgress>, Task<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>>> getFixedDocumentsAsync)
-        where TFixAllContext : IFixAllContext
-    {
-        // First, compute and apply the fixes.
-        var docIdToNewRootOrText = await getFixedDocumentsAsync(fixAllContext, progressTracker).ConfigureAwait(false);
+                // TODO: consider computing this in parallel.
+                var singleContextDocIdToNewRootOrText = await getFixedDocumentsAsync(fixAllContext, progressTracker).ConfigureAwait(false);
```

Comment:
>getFixedDocumentsAsync [](http://example.com/codeflow?start=62&length=22)

Is it possible that different calls to this return documents with the same id?

Reply:
no.  and i cna doc.  each  'fix all context' generally represents either an entirely different document, or an entirely different project.  so adding is entirely safe.

Reply:
ok, doc would definitely be appreciated. LGTM

Reply:
will add.

Reply:
Added in  00e08311077